Top-K optimization for ORDER BY + LIMIT

sreeharshar84 opened the pull request:
Issue:
------
A query containing ORDER BY and LIMIT is effectively the Top K (or Bottom K)
rows of the result. The current query execution sorts all the rows in the
table for the ORDER_BY operator and then applies the LIMIT operator.

This is wasteful in terms of both space and time. The inplace quick sort takes
up O(N) space and O(N^2) time (but O(N * logN) average time).

Solution:
---------
A better way is to use a heap and combine the ORDER BY and LIMIT operators.
This takes up less space O(K), where K is the LIMIT value (always less than N,
the number of rows) and O(N * logK) time.

Implementation:
---------------
1) New LogicalOperatorType and PhysicalOperatorType "ORDER_BY_LIMIT" has been
   introduced.
2) A new heap algorithm has been added to the ChunkCollection class.
3) When both ORDER BY AND LIMIT are present, the new LogicalOperatorType is
   introduced. This is then converted to the new PhysicalOperatorType for
   execution.
4) ORDER BY AND OFFSET (without LIMIT) will work as before. This optimzation
   doesn't apply for such queries.

sreeharshar84 committed:
Top-K optimization for ORDER BY + LIMIT

Issue:
------
A query containing ORDER BY and LIMIT is effectively the Top K (or Bottom K)
rows of the result. The current query execution sorts all the rows in the
table for the ORDER_BY operator and then applies the LIMIT operator.

This is wasteful in terms of both space and time. The inplace quick sort takes
up O(N) space and O(N^2) time (but O(N * logN) average time).

Solution:
---------
A better way is to use a heap and combine the ORDER BY and LIMIT operators.
This takes up less space O(K), where K is the LIMIT value (always less than N,
the number of rows) and O(N * logK) time.

Implementation:
---------------
1) New LogicalOperatorType and PhysicalOperatorType "ORDER_BY_LIMIT" has been
   introduced.
2) A new heap algorithm has been added to the ChunkCollection class.
3) When both ORDER BY AND LIMIT are present, the new LogicalOperatorType is
   introduced. This is then converted to the new PhysicalOperatorType for
   execution.
4) ORDER BY AND OFFSET (without LIMIT) will work as before. This optimzation
   doesn't apply for such queries.

hannesmuehleisen commented:
Thanks so much for your PR, looks good! I was wondering whether you considered doing the topn(order()) replace in the optimiser instead of the planner?

Mytherin commented on src/common/enums/logical_operator_type.cpp:
```diff
@@ -35,6 +35,8 @@ string LogicalOperatorToString(LogicalOperatorType type) {
 		return "LIMIT";
 	case LogicalOperatorType::ORDER_BY:
 		return "ORDER_BY";
+	case LogicalOperatorType::ORDER_BY_LIMIT:
```
Could you change this to "TOP_N" instead of "ORDER_BY_LIMIT"? Only a stylistic change but I think it conveys the meaning more clearly.

sreeharshar84 replied:
Agree. I will make this change.

Mytherin commented on src/execution/operator/order/CMakeLists.txt:
```diff
@@ -1,4 +1,7 @@
+add_library_unity(duckdb_operator_order
+                  OBJECT
+                  physical_order.cpp
+                  physical_order_limit.cpp)
```
Also please change this to `physical_top_n.cpp` and change the operators themselves to be called `PhysicalTopN` and `LogicalTopN`.

sreeharshar84 replied:
Agree.

Mytherin commented on src/planner/logical_plan/query_node/plan_query_node.cpp:
```diff
@@ -18,13 +19,29 @@ unique_ptr<LogicalOperator> LogicalPlanGenerator::VisitQueryNode(BoundQueryNode
 	}
 
-	if (node.orders.size() > 0) {
+	if (node.orders.size() > 0 && (node.limit >= 0 || node.offset >= 0)) {
```
As Hannes' has mentioned previously, converting from a limit and order by to a top N implementation should be done in the optimizer. It is not always beneficial to use a heap to do this, for example if N is very big the standard ORDER BY should perform better. As such, please leave the code in the planner as generating a limit and order, and then introduce an optimizer step in the Optimizer that looks for a limit with a child order by node and converts them into a top N node.

Mytherin commented on test/sql/simple/test_orderby.cpp:
```diff
@@ -153,6 +153,27 @@ TEST_CASE("Test ORDER BY exceptions", "[order]") {
+
+	// Top N queries that use ORDER_BY_LIMIT Optimization.
+	// Uses the new ORDER_BY_LIMIT optimization
+	result = con.Query("EXPLAIN SELECT a FROM test ORDER BY a LIMIT 1;");
```
I think it is not a good idea to rely on EXPLAIN output in tests, instead it would be better to test functionality. If you must test whether or not optimizers work as expected create an optimizer test that inspects the internal nodes instead. This is done in e.g. `test_cse_optimizer.cpp`

sreeharshar84 replied:
Agree. I will explore this possibility.

Mytherin reviewed (changes requested):
Thank you very much for your PR! I have added some comments, but the main two things are that (1) the conversion from limit + order by to top N should happen as part of the optimizer, as it is an optimization, and (2) the proper functioning of this optimization step should be tested by inspecting the internal nodes rather than looking at explain output as this is subject to change. Again, thanks for your contribution!

swingbit commented:
Mind that strictly speaking LIMIT + ORDER BY is not the same as TopN, as they both should only be concerned with the presentation of the result, not its computation.
I agree that in a simple case such as `SELECT .... ORDER BY x LIMIT 10` the optimizer can decide to make the two coincide.
I don't remember if nowadays standard SQL allows those two clauses in subqueries, it certainly didn't a few years back. If that is still the case, then it becomes a problem to, for example, ask for the TopN results (according to some criteria) and present them sorted in a different way.

If you like to dig into some old stuff for inspiration ;) : https://www.researchgate.net/publication/2318261_Optimization_and_Evaluation_of_Generalized_Top_Queries

sreeharshar84 committed:
Address the following issues:
1) Rename files physical_order_limit.cpp and logical_order_limit.cpp
   to physical_top_n.cpp and logical_top_n.cpp.
2) Rename "ORDER_BY_LIMIT" to "TOP_N" and "PhysicalOrderAndLimit" and
   "LogicalOrderAndLimit" to "PhysicalTopN" and "LogicalTopN".
3) A new optimizer step has been introduced to check for the presence
   of a LIMIT followed by an ORDER BY and replace them with a TOP N
   operator. This step was previously in the planner phase.
4) A new test_top_n.cpp has been added to test the optimizer phase.

Mytherin commented:
Thanks for the changes, it is looking much better! Playing around with it a bit, I found a segmentation fault if the OFFSET is bigger than the result set (rather than the expected result of an empty set):

```sql
CREATE TABLE integers(i INTEGER);
INSERT INTO integers VALUES (1), (2), (3), (4), (5);
SELECT * FROM integers ORDER BY i LIMIT 1 OFFSET 100;
```

Could you take a look at that problem and create a test case for that? Thanks!

sreeharshar84 committed:
Issue: When OFFSET value is greater than the total number of rows,
it results in a crash with the TOP N query.

Fix: In this case the heap size will be 0. There is no need for
further TOP N processing. Related tests have been added.

Mytherin commented:
Looks great now, thanks for your contribution!
